# views/map_page.py
import asyncio
import csv
import datetime
from PyQt5.QtWidgets import QWidget, QVBoxLayout
from PyQt5.QtCore import QUrl, QTimer
from PyQt5.QtWebEngineWidgets import QWebEngineView
import joblib
from receive_ble import scan_once
from trilateration import rssi_to_distance, trilaterate
from kalman import KalmanFilter2D
from ui.map_view import Ui_map_view
import os

class MapPage(QWidget):
    def __init__(self, main_window):
        super().__init__()
        self.ui = Ui_map_view()
        self.ui.setupUi(self)
        self.main_window = main_window

        # T·∫°o tr√¨nh duy·ªát map
        self.map_view = QWebEngineView()
        html_path = os.path.abspath("resource/map.html")
        self.map_view.load(QUrl.fromLocalFile(html_path))

        # Kh·ªüi t·∫°o Kalman Filter
        self.kalman_filter = KalmanFilter2D()

        layout = QVBoxLayout()
        layout.setContentsMargins(0, 0, 0, 0)
        self.map_view.setMinimumSize(500, 500)

        layout.addWidget(self.map_view)
        self.ui.map_container.setLayout(layout)

        # K·∫øt n·ªëi n√∫t ƒëi·ªÅu h∆∞·ªõng
        self.ui.shop_btn.clicked.connect(lambda: self.main_window.navigate_to("shop"))
        self.ui.list_btn.clicked.connect(lambda: self.main_window.navigate_to("cart"))

        # K·∫øt n·ªëi n√∫t ch·ªâ ƒë∆∞·ªùng
        self.ui.nagative_btn.clicked.connect(self.navigate_all_items)

        self.ui.scan_btn.toggled.connect(self.toggle_ble_scan)

        self.path_log_file = open("data/path_log.csv", "w", newline="")
        self.path_writer = csv.writer(self.path_log_file)
        self.path_writer.writerow(["timestamp", "x", "y"])
        # Load m√¥ h√¨nh
        self.model = joblib.load("model/ble_model.pkl")

        # Thi·∫øt l·∫≠p timer ƒë·ªÉ qu√©t BLE m·ªói 1 gi√¢y
        self.timer = QTimer(self)
        self.timer.timeout.connect(self.scan_and_update_position)

        self.beacons_pos = {
            "0-0": (0, 0),
            "0-30": (0, 30),
            "15-30": (15, 30)
        }
    def toggle_ble_scan(self, checked):
        if checked:
            self.ui.scan_btn.setText("üîÑ Scanning...")
            self.timer.start(1000)
            print("‚ñ∂Ô∏è B·∫Øt ƒë·∫ßu qu√©t BLE")
        else:
            self.ui.scan_btn.setText("‚ñ∂Ô∏è Start Scan")
            self.timer.stop()
            print("‚è∏Ô∏è D·ª´ng qu√©t BLE")
            
    def scan_and_update_position(self):
        try:
            rssi_dict = asyncio.run(scan_once())
            print("RSSI:", rssi_dict)

            # ---- D·ª± ƒëo√°n t·ª´ model ML ----
            input_vector = [rssi_dict.get("0-0", -100),
                            rssi_dict.get("0-30", -100),
                            rssi_dict.get("15-30", -100)]
            x_ml, y_ml = self.model.predict([input_vector])[0]

            # ---- T√≠nh kho·∫£ng c√°ch t·ª´ RSSI ----
            d1 = rssi_to_distance(rssi_dict["0-0"])
            d2 = rssi_to_distance(rssi_dict["0-30"])
            d3 = rssi_to_distance(rssi_dict["15-30"])

            # ---- D·ª± ƒëo√°n t·ª´ trilateration ----
            x_tri, y_tri = trilaterate(self.beacons_pos["0-0"], d1,
                                        self.beacons_pos["0-30"], d2,
                                        self.beacons_pos["15-30"], d3)

            # ---- Trung b√¨nh 2 k·∫øt qu·∫£ ho·∫∑c ch·ªçn k·∫øt qu·∫£ tin c·∫≠y h∆°n ----
            x = (x_ml + x_tri) / 2
            y = (y_ml + y_tri) / 2

            print(f"ML: ({x_ml:.1f}, {y_ml:.1f}) | Trilateration: ({x_tri:.1f}, {y_tri:.1f}) ‚Üí Combined: ({x:.1f}, {y:.1f})")
            # üß† L·ªçc Kalman
            self.kalman.predict()
            self.kalman.update([x, y])
            x_kf, y_kf = self.kalman.get_position()
            # ---- C·∫≠p nh·∫≠t marker tr√™n map ----
            self.map_view.page().runJavaScript(f"updateMarker({x_kf}, {y_kf})")
            timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            self.path_writer.writerow([timestamp, round(x_kf, 2), round(y_kf, 2)])
            self.path_log_file.flush()  # ƒë·∫£m b·∫£o ghi ngay
        except Exception as e:
            print("‚ùå L·ªói d·ª± ƒëo√°n v·ªã tr√≠:", e)
            
    def navigate_all_items(self):
        # L·∫•y danh s√°ch k·ªá t·ª´ b·∫£ng s·∫£n ph·∫©m
        aisle_list = []
        for row in range(self.ui.tableWidget.rowCount()):
            aisle_item = self.ui.tableWidget.item(row, 1)
            if aisle_item:
                aisle = aisle_item.text()
                if aisle and aisle not in aisle_list:
                    aisle_list.append(aisle)

        if aisle_list:
            js_array = str(aisle_list).replace("'", '"')  # Chuy·ªÉn sang m·∫£ng JavaScript
            self.map_view.page().runJavaScript(f'navigateAll({js_array})')
    def closeEvent(self, event):
        if hasattr(self, "data/path_log_file"):
            self.path_log_file.close()
        event.accept()